pep-0573.rst

```diff
@@ -0,0 +1,502 @@
+
+    * it is markedly slower than directly accessing C-level process-global state
+    * there is still some inherent reliance on process global state that means it still doesn't reliably handle module reloading
+
+It's also the case that when looking up a C-level struct such as module state, supplying
+an unexpected object layout can crash the interpreter, so it's significantly more important to ensure that extension
+methods receive the kind of object they expect.
+
+Proposal
+========
+
+Currently, a bound extension method (``PyCFunction`` or ``PyCFunctionWithKeywords``) receives only
+``self``, and (if applicable) the supplied positional and keyword arguments. 
+
+While module-level extension functions already receive access to the defining module object via their
+``self`` argument, methods of extension types don't have that luxury: they receive the bound instance
+via ``self``, and hence have no direct access to the defining class or the module level state.
+
+The additional module level context described above can be made available with two changes.
+Both additions are optional; extension authors need to opt in to start
+using them:
+
+    * Add a pointer to the module to heap type objects.
+
+    * Pass the defining class to the underlying C function.
+
+      The defining class is readily available at the time built-in
+      method object (``PyCFunctionObject``) is created, so it can be stored
+      in a new struct that extends ``PyCFunctionObject``.
+
+The module state can then be retrieved from the module object via
+``PyModule_GetState``.
+
+Note that this proposal implies that any type whose method needs to access
+per-module state must be a heap type, rather than a static type.
+
+This is necessary to support loading multiple module objects from a single
+extension: a static type, as a C-level global, has no information about
+which module it belongs to.
+
+
+Slot methods
+------------
+
+The above changes don't cover slot methods, such as ``tp_iter`` or ``nb_add``.
+
+The problem with slot methods is that their C API is fixed, so we can't
+simply add a new argument to pass in the defining class.
+Two possible solutions have been proposed to this problem:
+
+    * Look up the class through walking the MRO.
+      This is potentially expensive, but will be useful if performance is not
+      a problem (such as when raising a module-level exception).
+    * Storing a pointer to the defining class of each slot in a separate table,
+      ``__typeslots__`` [#typeslots-mail]_.  This is technically feasible and fast,
+      but quite invasive.
+
+Due to the invasiveness of the latter approach, this PEP proposes adding a MRO walking helper for use in slot method implementations, deferring the more complex alternative as a potential future optimisation.
```

Comment:
Modules affected by this concern also have the option of using thread local state or context variables, or else defining their own reload-friendly lookup caching scheme.